WordPress/Classes/Utility/Media/ImageDownloader.swift

```diff
@@ -115,32 +115,22 @@ actor ImageDownloader {
 
     private func data(for request: URLRequest, options: ImageRequestOptions) async throws -> Data {
         let requestKey = request.urlRequest?.url?.absoluteString ?? ""
-        let task = tasks[requestKey] ?? ImageDataTask(key: requestKey, Task {
-            try await self._data(for: request, options: options, key: requestKey)
-        })
-        task.downloader = self
 
-        let subscriptionID = UUID()
-        task.subscriptions.insert(subscriptionID)
-        tasks[requestKey] = task
-
-        return try await task.getData(subscriptionID: subscriptionID)
-    }
-
-    fileprivate nonisolated func unsubscribe(_ subscriptionID: UUID, key: String) {
-        Task {
-            await _unsubscribe(subscriptionID, key: key)
+        let theTask: Task<Data, any Error>
+        if let task = tasks[requestKey] {
+            theTask = task
+        } else {
+            theTask = Task {
+                try await self._data(for: request, options: options, key: requestKey)
+            }
+            tasks[requestKey] = theTask
         }
-    }
 
-    private func _unsubscribe(_ subscriptionID: UUID, key: String) {
-        guard let task = tasks[key],
```

Comment:
This version implements coalescing of tasks for the same URLs, which is especially useful for prefetching. If there is an outstanding task, it adds a subscriber to the existing task. The task gets canceled only when there are no subscribers left. It would ideally also reuse decompression/resizing, but it's a bit more challenging to implement, especially if you want to reuse these separate from the downloads.

**Background**: `ImageDownloader` is an existing class that I updated a while ago by adding background decompression, resizing, coalescing, and stuff like that. I also removed a few other systems. The downloader is currently the primary API for fetching images.

Reply:
Ah, of course. I don't know what I was thinking... The HTTP task is only canceled if all subscribers are canceled.